src/osdc/Objecter.cc

```diff
@@ -2316,6 +2319,52 @@ void Objecter::_op_submit(Op *op, shunique_lock<ceph::shared_mutex>& sul, ceph_t
   bool check_for_latest_map = _calc_target(&op->target, nullptr)
     == RECALC_OP_TARGET_POOL_DNE;
 
+  if (op->tid == 0)
+    op->tid = ++last_tid;
+  if (objecter_fastfail_timeout  > timespan(0)) {
+    unique_lock fl(ffpgs.lock);
+    if (ffpgs.pgs.find(op->target.pgid) != ffpgs.pgs.end()) {
+      fl.unlock();
+      ldout(cct, 10) << "fastfail pg: " << op->target.pgid << " to list" << dendl;
+      // before cancelling check if the pg is still inactive
+      // TODO: find a better way to determine if a pg is inactive
```

Comment:
what is going on here with TODO?

Reply:
@mattbenjamin I don't know if there is a better way to determine if the pg is inactive, currently I'm checking if the active size is smaller than min_size then it will enter the fastfail phase
@jdurgin @neha-ojha Is there a better way to check if pg is inactive?

Reply:
@jdurgin We already have the timeout option on osd op (rados_osd_op_timeout which doesn't have the fastfail itself but it timeouts requests).
Using inactive pg will miss some of the reasons, but include all is not good either (for example I'm not sure that we want to cancel op for busy HDD, slow response and we will not be able to cancel the fastfail if something works again) .
What is the condition to enter to fastfail, do you think we should use here?

Reply:
rados_osd_op_timeout is a new timeout for each op, whereas fastfail continues to fail new ops immediately until exiting the fastfail state. That is very different behavior.

We discussed a timing based approach, and that's what you'd described in a prior doc as well, so I'm confused why active/inactive is being treated specially. It's a very small set of cases that the client can tell a pg is inactive. There are many (e.g. waiting for an osd with the most up-to-date data to come back with > min_size osds) that clients don't have enough information to understand.

A general timeout handles these kinds of cases well. If RGW is going to handle errors well, I don't see why it would want to treat slowness due to waiting for an inactive pg that thet client can't detect any differently from a busy HDD - e.g. if an op will take 60s to complete, RGW could make better use of its memory and cpu resources during that time - it's essentially a backoff mechanism.

Reply:
This comment hasn't been addressed

Reply:
whoops I was just looking at the first commit. The 2nd commit addresses this well and can be squashed into the first. Thanks, and apologies it took so long to get back to this.